networkit/cpp/graph/GraphBuilder.cpp

```diff
@@ -12,86 +12,163 @@
+GraphBuilder::GraphBuilder(count n, bool weighted, bool directed, bool useWholeEdges)
+    : n(n), selfloops(0), weighted(weighted), directed(directed), useWholeEdges(useWholeEdges) {
+    // pre-allocating the adjacency vectors
+    int max_threads = omp_get_max_threads();
+    outEdgesPerThread.resize(max_threads);
+    inEdgesPerThread.resize(max_threads);
+    for (int i = 0; i < max_threads; i++) {
+        outEdgesPerThread[i].resize(max_threads);
+        inEdgesPerThread[i].resize(max_threads);
+    }
+    if (weighted) {
+        outEdgeWeightsPerThread.resize(max_threads);
+        inEdgeWeightsPerThread.resize(max_threads);
+        for (int i = 0; i < max_threads; i++) {
+            outEdgeWeightsPerThread[i].resize(max_threads);
+            inEdgeWeightsPerThread[i].resize(max_threads);
+        }
+    }
+}
 
 void GraphBuilder::reset(count n) {
     this->n = n;
     selfloops = 0;
-    outEdges.assign(n, std::vector<node>{});
-    outEdgeWeights.assign(isWeighted() ? n : 0, std::vector<edgeweight>{}),
-        inEdges.assign(isDirected() ? n : 0, std::vector<node>{}),
-        inEdgeWeights.assign((isDirected() && isWeighted()) ? n : 0, std::vector<edgeweight>{});
+
+    int max_threads = omp_get_max_threads();
+    outEdgesPerThread.assign(max_threads, std::vector<std::vector<HalfEdge>>{});
+    outEdgeWeightsPerThread.assign(isWeighted() ? max_threads : 0,
+                                   std::vector<std::vector<edgeweight>>{}),
+        inEdgesPerThread.assign(isDirected() ? max_threads : 0,
+                                std::vector<std::vector<HalfEdge>>{}),
+        inEdgeWeightsPerThread.assign((isDirected() && isWeighted()) ? max_threads : 0,
+                                      std::vector<std::vector<edgeweight>>{});
+    outEdgesPerThread.resize(max_threads);
+    for (int i = 0; i < max_threads; i++) {
+        outEdgesPerThread[i].resize(max_threads);
+    }
+    if (weighted) {
+        outEdgeWeightsPerThread.resize(max_threads);
+        for (int i = 0; i < max_threads; i++) {
+            outEdgeWeightsPerThread[i].resize(max_threads);
+        }
+    }
+    inEdgesPerThread.resize(max_threads);
+    if (weighted)
+        inEdgeWeightsPerThread.resize(max_threads);
+    for (int i = 0; i < max_threads; i++) {
+        inEdgesPerThread[i].resize(max_threads);
+        if (weighted)
+            inEdgeWeightsPerThread[i].resize(max_threads);
+    }
 }
 
-index GraphBuilder::indexInOutEdgeArray(node u, node v) const {
-    for (index i = 0; i < outEdges[u].size(); i++) {
-        node x = outEdges[u][i];
-        if (x == v) {
+index GraphBuilder::indexInOutEdgeArrayPerThread(node u, node v) const {
+    int max_threads = omp_get_max_threads();
+    int thread_num = omp_get_thread_num();
+    auto &edges = outEdgesPerThread[thread_num][u % max_threads];
+    for (index i = 0; i < edges.size(); ++i) {
+        if (edges[i].source == u && edges[i].destination == v)
             return i;
-        }
     }
```

Comment:
With https://en.cppreference.com/w/cpp/algorithm/all_any_none_of you can use something like:
```cpp
if (std::any_of(edges.begin(), edges.end(), [u, v](const HalfEdge& halfEdge) {
    return halfEdge.source == u && halfEdge.destination == v;
  }) {
    return i; 
} else {
  return none;
}

```

Reply:
This does not return the index which we are looking for. Besides that, where is the advantage ?

Reply:
I also can't see how we extract `i` with `std::any_of`. However, better define the result before the loop (init with `none`). If the index is found, set the result variable and break. This makes the code more readable.